sql/database.go

```diff
@@ -358,19 +612,59 @@ func (db *sqliteDatabase) getTx(ctx context.Context, initstmt string) (*sqliteTx
 	return tx, nil
 }
 
-func (db *sqliteDatabase) withTx(ctx context.Context, initstmt string, exec func(Transaction) error) error {
+func (db *sqliteDatabase) withTx(ctx context.Context, initstmt string, exec func(Transaction) error) (err error) {
 	tx, err := db.getTx(ctx, initstmt)
 	if err != nil {
 		return err
 	}
-	defer tx.Release()
+	defer func() {
+		if rErr := tx.Release(); rErr != nil && err == nil {
+			err = fmt.Errorf("release tx: %w", rErr)
+		}
+	}()
 	if err := exec(tx); err != nil {
 		tx.queryCache.ClearCache()
 		return err
 	}
 	return tx.Commit()
 }
 
+func (db *sqliteDatabase) startExclusive() error {
+	conn := db.getConn(context.Background())
+	if conn == nil {
+		return ErrNoConnection
+	}
+	// We don't need to wait for long if the database is busy
+	conn.SetBusyTimeout(1 * time.Millisecond)
+	// From SQLite docs:
+	// When the locking-mode is set to EXCLUSIVE, the database connection
+	// never releases file-locks. The first time the database is read in
+	// EXCLUSIVE mode, a shared lock is obtained and held. The first time the
+	// database is written, an exclusive lock is obtained and held.
+	if _, err := exec(conn, "PRAGMA locking_mode=EXCLUSIVE", nil, nil); err != nil {
+		db.pool.Put(conn)
```

Comment:
these are copied around and can just go into a `defer` statement after `L636`

Reply:
Fixed